File: app/src/main/java/com/batodev/arrows/engine/TransformationState.kt
package com.batodev.arrows.engine

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.geometry.Offset
import com.batodev.arrows.GameConstants

class TransformationState {
    var scale by mutableFloatStateOf(1f)
    var offsetX by mutableFloatStateOf(0f)
    var offsetY by mutableFloatStateOf(0f)

    fun reset() {
        scale = 1f
        offsetX = 0f
        offsetY = 0f
    }

    fun transform(pan: Offset, zoom: Float) {
        scale = (scale * zoom).coerceIn(GameConstants.MIN_SCALE, GameConstants.MAX_SCALE)
        offsetX += pan.x
        offsetY += pan.y
    }
}

File: flutter_arrows/lib/v3/modules/game/application/services/transformation_state.dart
import 'dart:math';
import 'package:flutter/material.dart';

import '../models/board_point.dart';
import '../models/game_constants.dart';

class TransformationState {
  double scale = GameConstants.defaultScale;
  double offsetX = 0.0;
  double offsetY = 0.0;

  void reset() {
    scale = GameConstants.defaultScale;
    offsetX = 0.0;
    offsetY = 0.0;
  }

  void transform(Offset pan, double zoom) {
    scale = (scale * zoom).clamp(GameConstants.minScale, GameConstants.maxScale);
    offsetX += pan.dx;
    offsetY += pan.dy;
  }
}
